src/main/java/com/mraof/minestuck/world/gen/structure/FrogTemplePiece.java

```diff
@@ -264,38 +265,21 @@ private Selector()
 		@Override
 		public void next(Random rand, int x, int y, int z, boolean wall)
 		{
-			int randomBlock = rand.nextInt(14);
-			if(randomBlock == 12 || randomBlock == 13)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_FROG.get().defaultBlockState();
-			} else if(randomBlock == 10 || randomBlock == 11)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_TURTLE.get().defaultBlockState();
-			} else if(randomBlock == 8 || randomBlock == 9)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_SKAIA.get().defaultBlockState();
-			} else if(randomBlock == 6 || randomBlock == 7)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_LOTUS.get().defaultBlockState();
-			} else if(randomBlock == 5)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_IGUANA_LEFT.get().defaultBlockState();
-			} else if(randomBlock == 4)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_IGUANA_RIGHT.get().defaultBlockState();
-			} else if(randomBlock == 3)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_NAK_LEFT.get().defaultBlockState();
-			} else if(randomBlock == 2)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_NAK_RIGHT.get().defaultBlockState();
-			} else if(randomBlock == 1)
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_SALAMANDER_LEFT.get().defaultBlockState();
-			} else
-			{
-				this.next = MSBlocks.GREEN_STONE_BRICK_SALAMANDER_RIGHT.get().defaultBlockState();
-			}
+			List<WeightedEntry.Wrapper<Block>> weightedBlockList = Lists.newArrayList();
+			
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_FROG.get(), 1)); //only one frog hieroglyph in order to have a component more difficult to find
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_TURTLE.get(), 36));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_SKAIA.get(), 36));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_LOTUS.get(), 36));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_IGUANA_LEFT.get(), 16));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_IGUANA_RIGHT.get(), 16));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_NAK_LEFT.get(), 16));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_NAK_RIGHT.get(), 16));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_SALAMANDER_LEFT.get(), 16));
+			weightedBlockList.add(WeightedEntry.wrap(MSBlocks.GREEN_STONE_BRICK_SALAMANDER_RIGHT.get(), 16));
```

Comment:
Since this function is being called plenty of times, I'd suggest building the list in the constructor instead. That way it only happens once per structure, instead of once per block.

Reply:
If you do so, you could also store the total weight from `WeightedRandom.getTotalWeight()` and pass it into `WeightedRandom.getRandomItem()` as well, since that'd also be the same for every block.

Reply:
I got it working to what I believe is what you recommended but as before please check to make sure it was an efficient approach